File: python/pentomino.py
#!/bin/python

import sys
from typing import List
from pieces import PIECES

def get_pieces() -> List:
	"""
	PIECES is a dict associating to each piece the list of its permutations.
	"""
	pieces = []

	if len(sys.argv) != 2:
		print("Usage: {} <pieces>".format(sys.argv[0]), file=sys.stderr)
		exit(1)

	for c in sys.argv[1]:
		piece = PIECES.get(c)
		if piece is None:
			print("Error: invalid piece found. Valid pieces are FILNPTUVWXYZ.", file=sys.stderr)
			exit(1)
		pieces.append(piece)
	return pieces

def print_board(board) -> None:
	for y in board:
		for x in y:
			print(x, end='')
		print('')

def can_be_placed(piece, board, y, x) -> bool:
	try:
		for i in range(len(piece)):
			for j in range(len(piece[i])):
				if piece[i][j] != '0' and board[y + i][x + j] != '0':
					return False
	except IndexError:
		return False
	return True

def fill_board(piece, board, y, x, place) -> None:
	"""
	This function both places and removes pieces.
	"""
	for i in range(len(piece)):
		for j in range(len(piece[i])):
			if piece[i][j] != '0':
				board[y + i][x + j] = piece[i][j] if place else '0'

def solve(pieces, current, board) -> bool:
	"""
	This is a simple backtracking. For loops are slow in python, but I think
	this is the simplest implementation.
	Each piece is placed wherever possible, and calls for the next one.
	If all pieces are placed, this is good. Else, pieces are removed and
	another position / piece rotation is tried.
	"""
	rotations = pieces[current]
	for piece in rotations:
		for y in range(len(board)):
			for x in range(len(board[y])):
				if can_be_placed(piece, board, y, x):
					fill_board(piece, board, y, x, True)
					if current == len(pieces) - 1:
						return True
					elif solve(pieces, current + 1, board) is True:
						return True
					else:
						fill_board(piece, board, y, x, False)
	return False

def main() -> None:
	"""
	The pentomino setup is a grid of size 5xN with n the number of pieces.
	"""
	pieces = get_pieces()
	board = [['0' for i in range(len(sys.argv[1]))] for i in range(5)]

	if solve(pieces, 0, board) is True:
		print_board(board)
	else:
		print('This setup seems impossible...')

if __name__ == "__main__":
	main()

File: C/pentomino.c
#include "pentomino.h"

static void	*safe_alloc(size_t n)
{
	void	*ret = malloc(n);
	if (ret == NULL)
	{
		dprintf(2, "Error: Couldn't allocate memory\n");
		exit(EXIT_FAILURE);
	}
	return ret;
}

static int get_rotations(t_piece *piece, char c)
{
	switch (c)
	{
		case 'F':
			piece->rotations = (char**)F_ROTATIONS;
			piece->nb_rotations = COUNT_OF(F_ROTATIONS);
			return 1;
		case 'I':
			piece->rotations = (char**)I_ROTATIONS;
			piece->nb_rotations = COUNT_OF(I_ROTATIONS);
			return 1;
		case 'L':
			piece->rotations = (char**)L_ROTATIONS;
			piece->nb_rotations = COUNT_OF(L_ROTATIONS);
			return 1;
		case 'N':
			piece->rotations = (char**)N_ROTATIONS;
			piece->nb_rotations = COUNT_OF(N_ROTATIONS);
			return 1;
		case 'P':
			piece->rotations = (char**)P_ROTATIONS;
			piece->nb_rotations = COUNT_OF(P_ROTATIONS);
			return 1;
		case 'T':
			piece->rotations = (char**)T_ROTATIONS;
			piece->nb_rotations = COUNT_OF(T_ROTATIONS);
			return 1;
		case 'U':
			piece->rotations = (char**)U_ROTATIONS;
			piece->nb_rotations = COUNT_OF(U_ROTATIONS);
			return 1;
		case 'V':
			piece->rotations = (char**)V_ROTATIONS;
			piece->nb_rotations = COUNT_OF(V_ROTATIONS);
			return 1;
		case 'W':
			piece->rotations = (char**)W_ROTATIONS;
			piece->nb_rotations = COUNT_OF(W_ROTATIONS);
			return 1;
		case 'X':
			piece->rotations = (char**)X_ROTATIONS;
			piece->nb_rotations = COUNT_OF(X_ROTATIONS);
			return 1;
		case 'Y':
			piece->rotations = (char**)Y_ROTATIONS;
			piece->nb_rotations = COUNT_OF(Y_ROTATIONS);
			return 1;
		case 'Z':
			piece->rotations = (char**)Z_ROTATIONS;
			piece->nb_rotations = COUNT_OF(Z_ROTATIONS);
			return 1;
		default:
			return 0;
	}
}

static t_piece *get_pieces(size_t nb_pieces, char *av)
{
	t_piece *pieces = safe_alloc(sizeof(t_piece) * nb_pieces);

	for(size_t i = 0; i < nb_pieces; i++)
	{
		if (get_rotations(&pieces[i], av[i]) == 0)
		{
			dprintf(2, "Error: invalid piece found. Valid pieces are FILNPTUVWXYZ.\n");
			free(pieces);
			exit(EXIT_FAILURE);
		}
	}
	return pieces;
}

static char **get_board(size_t nb_pieces)
{
	char **board = safe_alloc(sizeof(char*) * 5);

	for(size_t i = 0; i < 5; i++)
	{
		board[i] = safe_alloc(nb_pieces + 1);
		memset(board[i], '0', nb_pieces);
		board[i][nb_pieces] = '\0';
	}
	return board;
}

static void print_board(char **board)
{
	for (size_t i = 0; i < 5; i++)
		printf("%s\n", board[i]);
}

static int can_be_placed(char **piece, char **board, size_t x, size_t y, size_t board_size)
{
	size_t	part = 0;

	for (size_t i = 0; i < 5; i++)
		for (size_t j = 0; j < 5; j++)
		{
			if (piece[i][j] != '0')
			{
				if (x + i >= 5 || y + j >= board_size
						|| board[x + i][y + j] != '0')
					return 0;
				part++;
				if(part == 5)
					return 1;
			}
		}
	return 1;
}

static void fill_board(char **piece, char **board, size_t x, size_t y, int place)
{
	size_t part = 0;

	for(size_t i = 0; i < 5; i++)
		for (size_t j = 0; j < 5; j++)
		{
			if (piece[i][j] != '0')
			{
				board[x + i][y + j] = place ? piece[i][j] : '0';
				part++;
			}
			if (part == 5)
				return;
		}
}

static int solve(t_piece *pieces, size_t current, char **board, size_t nb_pieces)
{
	char **rotation;

	for(size_t i = 0; i < pieces[current].nb_rotations; i++)
	{
		rotation = &pieces[current].rotations[i * 5];
		for (size_t x = 0; x < 5; x++)
		{
			for (size_t y = 0; y < nb_pieces; y++)
			{
				if (can_be_placed(rotation, board, x, y, nb_pieces))
				{
					fill_board(rotation, board, x, y, 1);
					if (current == nb_pieces - 1)
						return 1;
					else if (solve(pieces, current + 1, board, nb_pieces))
						return 1;
					else
						fill_board(rotation, board, x, y, 0);
				}
			}
		}
	}
	return 0;
}

static void	free_resources(t_piece *pieces, char **board)
{
	for(size_t i = 0; i < 5; i++)
		free(board[i]);
	free(board);
	free(pieces);
}

int main(int ac, char **av)
{
	t_piece *pieces = NULL;
	char	**board = NULL;
	size_t nb_pieces;

	if (ac != 2)
	{
		dprintf(2, "Usage: %s <pieces>\n", av[0]);
		return EXIT_FAILURE;
	}
	nb_pieces = strlen(av[1]);
	pieces = get_pieces(nb_pieces, av[1]);
	board = get_board(nb_pieces);
	if (solve(pieces, 0, board, nb_pieces))
		print_board(board);
	else
		printf("This setup seems impossible...\n");
	free_resources(pieces, board);
	return EXIT_SUCCESS;
}
